File: project/frontend-services/libs/smart-irrigation/mapper/src/lib/mappers/DeviceMapper.ts
import {DeviceDTO, DeviceTypeDTO} from "@frontend-services/smart-irrigation/dto";
import {Device, DeviceId, DeviceName, DeviceType, RecordEntry} from "@frontend-services/smart-irrigation/model";

export class DeviceMapper {
  static dtoToModel(dto: DeviceDTO): Device {
    const id = new DeviceId(dto.id);
    const name = new DeviceName(dto.name);
    const records = dto.records.map(d => new RecordEntry(d.label, d.content));
    switch (dto.type) {
      case DeviceTypeDTO.PARK_SENSOR:
        return new Device(id, DeviceType.PARK_SENSOR, name, records, dto.remoteControl, false);
      case DeviceTypeDTO.STOVE_SENSOR:
        return new Device(id, DeviceType.STOVE_SENSOR, name, records, dto.remoteControl, false);
      case DeviceTypeDTO.VALVE:
        return new Device(id, DeviceType.VALVE, name, records, dto.remoteControl, false);
    }
  }
}

File: project/backend-services/smart-irrigation-backend/infrastructure/persistence/postgres/src/main/java/pt/sensae/services/smart/irrigation/backend/infrastructure/persistence/postgres/mapper/device/DeviceMapper.java
package pt.sensae.services.smart.irrigation.backend.infrastructure.persistence.postgres.mapper.device;

import pt.sensae.services.smart.irrigation.backend.domain.exceptions.NotValidException;
import pt.sensae.services.smart.irrigation.backend.domain.model.business.device.Device;
import pt.sensae.services.smart.irrigation.backend.domain.model.business.device.DeviceId;
import pt.sensae.services.smart.irrigation.backend.domain.model.business.device.DeviceType;
import pt.sensae.services.smart.irrigation.backend.domain.model.business.device.ledger.DeviceLedger;
import pt.sensae.services.smart.irrigation.backend.domain.model.business.device.ledger.LedgerEntry;
import pt.sensae.services.smart.irrigation.backend.infrastructure.persistence.postgres.model.device.DevicePostgres;

import java.util.Set;
import java.util.UUID;

public class DeviceMapper {

    public static DevicePostgres modelToDao(Device model) {
        var type = switch (model.type()) {
            case PARK_SENSOR -> "park_sensor";
            case VALVE -> "valve";
            case STOVE_SENSOR -> "stove_sensor";
        };
        return new DevicePostgres(model.id().value().toString(), type);
    }

    public static Device daoToModel(DevicePostgres dao, Set<LedgerEntry> entry) {
        var deviceType = switch (dao.deviceType) {
            case "park_sensor" -> DeviceType.PARK_SENSOR;
            case "valve" -> DeviceType.VALVE;
            case "stove_sensor" -> DeviceType.STOVE_SENSOR;
            default -> throw new NotValidException("No Valid data packet found");
        };
        return new Device(DeviceId.of(UUID.fromString(dao.deviceId)), deviceType, new DeviceLedger(entry));
    }
}
